{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f751e6d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'HiC_scaffold_1': 88721917,\n",
       " 'HiC_scaffold_2': 79379215,\n",
       " 'HiC_scaffold_3': 74701305,\n",
       " 'HiC_scaffold_4': 73032783,\n",
       " 'HiC_scaffold_5': 70278957,\n",
       " 'HiC_scaffold_6': 70171953,\n",
       " 'HiC_scaffold_7': 69140750,\n",
       " 'HiC_scaffold_8': 63626884,\n",
       " 'HiC_scaffold_9': 62461566,\n",
       " 'HiC_scaffold_10': 62032692,\n",
       " 'HiC_scaffold_11': 60969504,\n",
       " 'HiC_scaffold_12': 60416524,\n",
       " 'HiC_scaffold_13': 59927272,\n",
       " 'HiC_scaffold_14': 59492018,\n",
       " 'HiC_scaffold_15': 54906828,\n",
       " 'HiC_scaffold_16': 53887023,\n",
       " 'HiC_scaffold_17': 51816475,\n",
       " 'HiC_scaffold_18': 45703695}"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from tqdm import tqdm\n",
    "import sys\n",
    "\n",
    "sys.path.append('/private/groups/shapirolab/brock/Software')\n",
    "from py3_functions import *\n",
    "\n",
    "from IPython.display import display\n",
    "\n",
    "!mkdir -p /private/groups/shapirolab/brock/mutation\n",
    "os.chdir('/private/groups/shapirolab/brock/mutation')\n",
    "!mkdir -p cmds logs\n",
    "\n",
    "### R stuff\n",
    "from rpy2 import rinterface\n",
    "#from jupyter_helpers import rpy2_autocompletion\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "samples = []\n",
    "with open('poplists/samples.txt') as infile:\n",
    "    for line in infile:\n",
    "        line = line.strip()\n",
    "        samples.append(line)\n",
    "infile.close()\n",
    "\n",
    "AB_sizes = {}\n",
    "with open('white_abalone.fasta.fai','r') as infile:\n",
    "    for line in infile:\n",
    "        chrom, size, total, b, c = line.split('\\t')\n",
    "        if int(size) > 1e6:\n",
    "            AB_sizes[chrom] = int(size)\n",
    "infile.close()\n",
    "AB_sizes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7f655e3",
   "metadata": {},
   "source": [
    "# Whatshap"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc3aa9b6-a93e-466c-b6d7-ff80a4d11b4c",
   "metadata": {},
   "source": [
    "**Warning**: attempting to use pedigree information for phasing with `whatshap phase --ped` will cause whatshapp to skip Mendelian violations, which are exactly the variants I want to phase. Currently can't find a workaround, so I phase without pedigree information to assist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e0c56bf6",
   "metadata": {
    "code_folding": [
     5
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/bin/bash\n",
      "#SBATCH -p long\n",
      "#SBATCH -t 120:00:00\n",
      "#SBATCH --mem=200000\n",
      "#SBATCH -n 1\n",
      "#SBATCH --array=1-1%1\n",
      "#SBATCH -e ./logs/HiC_scaffold_1.whatshap.e\n",
      "#SBATCH -o ./logs/HiC_scaffold_1.whatshap.o\n",
      "#SBATCH -c 10\n",
      "#SBATCH -N 1\n",
      "#SBATCH -J HiC_scaffold_1.whatshap\n",
      "\n",
      "\n",
      "ls cmds/HiC_scaffold_1:*whatshap.sh | sed 's/^/bash /g' | parallel --progress --jobs 10\n",
      "\n",
      "#!/bin/bash\n",
      "#SBATCH -p long\n",
      "#SBATCH -t 120:00:00\n",
      "#SBATCH --mem=200000\n",
      "#SBATCH -n 1\n",
      "#SBATCH --array=1-1%1\n",
      "#SBATCH -e ./logs/HiC_scaffold_2.whatshap.e\n",
      "#SBATCH -o ./logs/HiC_scaffold_2.whatshap.o\n",
      "#SBATCH -c 10\n",
      "#SBATCH -N 1\n",
      "#SBATCH -J HiC_scaffold_2.whatshap\n",
      "\n",
      "\n",
      "ls cmds/HiC_scaffold_2:*whatshap.sh | sed 's/^/bash /g' | parallel --progress --jobs 10\n",
      "\n",
      "#!/bin/bash\n",
      "#SBATCH -p long\n",
      "#SBATCH -t 120:00:00\n",
      "#SBATCH --mem=200000\n",
      "#SBATCH -n 1\n",
      "#SBATCH --array=1-1%1\n",
      "#SBATCH -e ./logs/HiC_scaffold_3.whatshap.e\n",
      "#SBATCH -o ./logs/HiC_scaffold_3.whatshap.o\n",
      "#SBATCH -c 10\n",
      "#SBATCH -N 1\n",
      "#SBATCH -J HiC_scaffold_3.whatshap\n",
      "\n",
      "\n",
      "ls cmds/HiC_scaffold_3:*whatshap.sh | sed 's/^/bash /g' | parallel --progress --jobs 10\n",
      "\n",
      "#!/bin/bash\n",
      "#SBATCH -p long\n",
      "#SBATCH -t 120:00:00\n",
      "#SBATCH --mem=200000\n",
      "#SBATCH -n 1\n",
      "#SBATCH --array=1-1%1\n",
      "#SBATCH -e ./logs/HiC_scaffold_4.whatshap.e\n",
      "#SBATCH -o ./logs/HiC_scaffold_4.whatshap.o\n",
      "#SBATCH -c 10\n",
      "#SBATCH -N 1\n",
      "#SBATCH -J HiC_scaffold_4.whatshap\n",
      "\n",
      "\n",
      "ls cmds/HiC_scaffold_4:*whatshap.sh | sed 's/^/bash /g' | parallel --progress --jobs 10\n",
      "\n",
      "#!/bin/bash\n",
      "#SBATCH -p long\n",
      "#SBATCH -t 120:00:00\n",
      "#SBATCH --mem=200000\n",
      "#SBATCH -n 1\n",
      "#SBATCH --array=1-1%1\n",
      "#SBATCH -e ./logs/HiC_scaffold_5.whatshap.e\n",
      "#SBATCH -o ./logs/HiC_scaffold_5.whatshap.o\n",
      "#SBATCH -c 10\n",
      "#SBATCH -N 1\n",
      "#SBATCH -J HiC_scaffold_5.whatshap\n",
      "\n",
      "\n",
      "ls cmds/HiC_scaffold_5:*whatshap.sh | sed 's/^/bash /g' | parallel --progress --jobs 10\n",
      "\n",
      "#!/bin/bash\n",
      "#SBATCH -p long\n",
      "#SBATCH -t 120:00:00\n",
      "#SBATCH --mem=200000\n",
      "#SBATCH -n 1\n",
      "#SBATCH --array=1-1%1\n",
      "#SBATCH -e ./logs/HiC_scaffold_6.whatshap.e\n",
      "#SBATCH -o ./logs/HiC_scaffold_6.whatshap.o\n",
      "#SBATCH -c 10\n",
      "#SBATCH -N 1\n",
      "#SBATCH -J HiC_scaffold_6.whatshap\n",
      "\n",
      "\n",
      "ls cmds/HiC_scaffold_6:*whatshap.sh | sed 's/^/bash /g' | parallel --progress --jobs 10\n",
      "\n",
      "#!/bin/bash\n",
      "#SBATCH -p long\n",
      "#SBATCH -t 120:00:00\n",
      "#SBATCH --mem=200000\n",
      "#SBATCH -n 1\n",
      "#SBATCH --array=1-1%1\n",
      "#SBATCH -e ./logs/HiC_scaffold_7.whatshap.e\n",
      "#SBATCH -o ./logs/HiC_scaffold_7.whatshap.o\n",
      "#SBATCH -c 10\n",
      "#SBATCH -N 1\n",
      "#SBATCH -J HiC_scaffold_7.whatshap\n",
      "\n",
      "\n",
      "ls cmds/HiC_scaffold_7:*whatshap.sh | sed 's/^/bash /g' | parallel --progress --jobs 10\n",
      "\n",
      "#!/bin/bash\n",
      "#SBATCH -p long\n",
      "#SBATCH -t 120:00:00\n",
      "#SBATCH --mem=200000\n",
      "#SBATCH -n 1\n",
      "#SBATCH --array=1-1%1\n",
      "#SBATCH -e ./logs/HiC_scaffold_8.whatshap.e\n",
      "#SBATCH -o ./logs/HiC_scaffold_8.whatshap.o\n",
      "#SBATCH -c 10\n",
      "#SBATCH -N 1\n",
      "#SBATCH -J HiC_scaffold_8.whatshap\n",
      "\n",
      "\n",
      "ls cmds/HiC_scaffold_8:*whatshap.sh | sed 's/^/bash /g' | parallel --progress --jobs 10\n",
      "\n",
      "#!/bin/bash\n",
      "#SBATCH -p long\n",
      "#SBATCH -t 120:00:00\n",
      "#SBATCH --mem=200000\n",
      "#SBATCH -n 1\n",
      "#SBATCH --array=1-1%1\n",
      "#SBATCH -e ./logs/HiC_scaffold_9.whatshap.e\n",
      "#SBATCH -o ./logs/HiC_scaffold_9.whatshap.o\n",
      "#SBATCH -c 10\n",
      "#SBATCH -N 1\n",
      "#SBATCH -J HiC_scaffold_9.whatshap\n",
      "\n",
      "\n",
      "ls cmds/HiC_scaffold_9:*whatshap.sh | sed 's/^/bash /g' | parallel --progress --jobs 10\n",
      "\n",
      "#!/bin/bash\n",
      "#SBATCH -p long\n",
      "#SBATCH -t 120:00:00\n",
      "#SBATCH --mem=200000\n",
      "#SBATCH -n 1\n",
      "#SBATCH --array=1-1%1\n",
      "#SBATCH -e ./logs/HiC_scaffold_10.whatshap.e\n",
      "#SBATCH -o ./logs/HiC_scaffold_10.whatshap.o\n",
      "#SBATCH -c 10\n",
      "#SBATCH -N 1\n",
      "#SBATCH -J HiC_scaffold_10.whatshap\n",
      "\n",
      "\n",
      "ls cmds/HiC_scaffold_10:*whatshap.sh | sed 's/^/bash /g' | parallel --progress --jobs 10\n",
      "\n",
      "#!/bin/bash\n",
      "#SBATCH -p long\n",
      "#SBATCH -t 120:00:00\n",
      "#SBATCH --mem=200000\n",
      "#SBATCH -n 1\n",
      "#SBATCH --array=1-1%1\n",
      "#SBATCH -e ./logs/HiC_scaffold_11.whatshap.e\n",
      "#SBATCH -o ./logs/HiC_scaffold_11.whatshap.o\n",
      "#SBATCH -c 10\n",
      "#SBATCH -N 1\n",
      "#SBATCH -J HiC_scaffold_11.whatshap\n",
      "\n",
      "\n",
      "ls cmds/HiC_scaffold_11:*whatshap.sh | sed 's/^/bash /g' | parallel --progress --jobs 10\n",
      "\n",
      "#!/bin/bash\n",
      "#SBATCH -p long\n",
      "#SBATCH -t 120:00:00\n",
      "#SBATCH --mem=200000\n",
      "#SBATCH -n 1\n",
      "#SBATCH --array=1-1%1\n",
      "#SBATCH -e ./logs/HiC_scaffold_12.whatshap.e\n",
      "#SBATCH -o ./logs/HiC_scaffold_12.whatshap.o\n",
      "#SBATCH -c 10\n",
      "#SBATCH -N 1\n",
      "#SBATCH -J HiC_scaffold_12.whatshap\n",
      "\n",
      "\n",
      "ls cmds/HiC_scaffold_12:*whatshap.sh | sed 's/^/bash /g' | parallel --progress --jobs 10\n",
      "\n",
      "#!/bin/bash\n",
      "#SBATCH -p long\n",
      "#SBATCH -t 120:00:00\n",
      "#SBATCH --mem=200000\n",
      "#SBATCH -n 1\n",
      "#SBATCH --array=1-1%1\n",
      "#SBATCH -e ./logs/HiC_scaffold_13.whatshap.e\n",
      "#SBATCH -o ./logs/HiC_scaffold_13.whatshap.o\n",
      "#SBATCH -c 10\n",
      "#SBATCH -N 1\n",
      "#SBATCH -J HiC_scaffold_13.whatshap\n",
      "\n",
      "\n",
      "ls cmds/HiC_scaffold_13:*whatshap.sh | sed 's/^/bash /g' | parallel --progress --jobs 10\n",
      "\n",
      "#!/bin/bash\n",
      "#SBATCH -p long\n",
      "#SBATCH -t 120:00:00\n",
      "#SBATCH --mem=200000\n",
      "#SBATCH -n 1\n",
      "#SBATCH --array=1-1%1\n",
      "#SBATCH -e ./logs/HiC_scaffold_14.whatshap.e\n",
      "#SBATCH -o ./logs/HiC_scaffold_14.whatshap.o\n",
      "#SBATCH -c 10\n",
      "#SBATCH -N 1\n",
      "#SBATCH -J HiC_scaffold_14.whatshap\n",
      "\n",
      "\n",
      "ls cmds/HiC_scaffold_14:*whatshap.sh | sed 's/^/bash /g' | parallel --progress --jobs 10\n",
      "\n",
      "#!/bin/bash\n",
      "#SBATCH -p long\n",
      "#SBATCH -t 120:00:00\n",
      "#SBATCH --mem=200000\n",
      "#SBATCH -n 1\n",
      "#SBATCH --array=1-1%1\n",
      "#SBATCH -e ./logs/HiC_scaffold_15.whatshap.e\n",
      "#SBATCH -o ./logs/HiC_scaffold_15.whatshap.o\n",
      "#SBATCH -c 10\n",
      "#SBATCH -N 1\n",
      "#SBATCH -J HiC_scaffold_15.whatshap\n",
      "\n",
      "\n",
      "ls cmds/HiC_scaffold_15:*whatshap.sh | sed 's/^/bash /g' | parallel --progress --jobs 10\n",
      "\n",
      "#!/bin/bash\n",
      "#SBATCH -p long\n",
      "#SBATCH -t 120:00:00\n",
      "#SBATCH --mem=200000\n",
      "#SBATCH -n 1\n",
      "#SBATCH --array=1-1%1\n",
      "#SBATCH -e ./logs/HiC_scaffold_16.whatshap.e\n",
      "#SBATCH -o ./logs/HiC_scaffold_16.whatshap.o\n",
      "#SBATCH -c 10\n",
      "#SBATCH -N 1\n",
      "#SBATCH -J HiC_scaffold_16.whatshap\n",
      "\n",
      "\n",
      "ls cmds/HiC_scaffold_16:*whatshap.sh | sed 's/^/bash /g' | parallel --progress --jobs 10\n",
      "\n",
      "#!/bin/bash\n",
      "#SBATCH -p long\n",
      "#SBATCH -t 120:00:00\n",
      "#SBATCH --mem=200000\n",
      "#SBATCH -n 1\n",
      "#SBATCH --array=1-1%1\n",
      "#SBATCH -e ./logs/HiC_scaffold_17.whatshap.e\n",
      "#SBATCH -o ./logs/HiC_scaffold_17.whatshap.o\n",
      "#SBATCH -c 10\n",
      "#SBATCH -N 1\n",
      "#SBATCH -J HiC_scaffold_17.whatshap\n",
      "\n",
      "\n",
      "ls cmds/HiC_scaffold_17:*whatshap.sh | sed 's/^/bash /g' | parallel --progress --jobs 10\n",
      "\n",
      "#!/bin/bash\n",
      "#SBATCH -p long\n",
      "#SBATCH -t 120:00:00\n",
      "#SBATCH --mem=200000\n",
      "#SBATCH -n 1\n",
      "#SBATCH --array=1-1%1\n",
      "#SBATCH -e ./logs/HiC_scaffold_18.whatshap.e\n",
      "#SBATCH -o ./logs/HiC_scaffold_18.whatshap.o\n",
      "#SBATCH -c 10\n",
      "#SBATCH -N 1\n",
      "#SBATCH -J HiC_scaffold_18.whatshap\n",
      "\n",
      "\n",
      "ls cmds/HiC_scaffold_18:*whatshap.sh | sed 's/^/bash /g' | parallel --progress --jobs 10\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!mkdir -p whatshap whatshap/bams/ whatshap/phased whatshap/vcfs\n",
    " \n",
    "    \n",
    "AB_sizes = {}\n",
    "with open('white_abalone.fasta.fai','r') as infile:\n",
    "    for line in infile:\n",
    "        chrom, size, total, b, c = line.split('\\t')\n",
    "        if int(size) > 1e6:\n",
    "            AB_sizes[chrom] = int(size)\n",
    "infile.close()\n",
    "AB_sizes\n",
    "\n",
    "\n",
    "#### This code creates a single slurm script PER CHROMOSOME ####\n",
    "#### but parallelizes within the slurm script using subregions ####\n",
    "#### and multithreading ####\n",
    "for CHROM, SIZE in AB_sizes.items():\n",
    "    outfile = open(f\"cmds/{CHROM}_whatshap_phase.txt\",'w')\n",
    "    ##\n",
    "    SIZE = int(SIZE)\n",
    "    for I in range(1,SIZE,5000000):\n",
    "        \n",
    "        ### REGION SETUP AND CHECKS ###\n",
    "        START = I\n",
    "        END = I + 5000000 - 1\n",
    "        if END > SIZE:\n",
    "            END = SIZE\n",
    "        if os.path.exists('whatshap/phased/%s:%s-%s.whatshap.vcf.gz' % (CHROM, START, END)): continue\n",
    "        REGION = '%s:%s-%s' % (CHROM, START, END)\n",
    "        baminput = ''\n",
    "        subfile = open(f\"cmds/{REGION}.whatshap.sh\", 'w')\n",
    "        \n",
    "        ### VCF CMD ###\n",
    "        cmd = f\"bcftools view -O z -r {REGION} vcfs/biallelic.vcf.gz > whatshap/vcfs/{REGION}.vcf.gz && tabix -f whatshap/vcfs/{REGION}.vcf.gz\\n\"\n",
    "        subfile.write(cmd)\n",
    "        \n",
    "        ### SUBBAMS CMDS ###\n",
    "        for SAMPLE in samples:\n",
    "            BAM = f\"aln/{SAMPLE}.mdup.bam\"\n",
    "            SUBBAM = f\"whatshap/bams/{SAMPLE}.{REGION}.bam\"\n",
    "            if not os.path.exists(SUBBAM + '.bai'):\n",
    "                samcmd = f\"samtools view -b -1 -M {BAM} {REGION} > {SUBBAM} && samtools index -b {SUBBAM}\\n\"\n",
    "                subfile.write(samcmd)\n",
    "            baminput += '%s ' % SUBBAM\n",
    "        \n",
    "        ### PHASE CMD ###\n",
    "        phasecmd = ('''\\nwhatshap phase '''\n",
    "                    '''--ped poplists/plink.ped '''\n",
    "                    '''whatshap/vcfs/{REGION}.vcf.gz '''\n",
    "                    '''{baminput} '''\n",
    "                    '''-o whatshap/phased/{REGION}.whatshap.vcf '''\n",
    "                    '''--reference=white_abalone.fasta '''\n",
    "                    #'''--ignore-read-groups '''\n",
    "                    '''--chromosome {CHROM} '''\n",
    "                    '''&& bgzip -f whatshap/phased/{REGION}.whatshap.vcf && tabix -f whatshap/phased/{REGION}.whatshap.vcf.gz &> logs/{REGION}.whatshap.log\\n''').format(baminput = baminput, REGION = REGION, CHROM = CHROM)\n",
    "        subfile.write(phasecmd)\n",
    "        \n",
    "        ### CLEANUP CMD ###\n",
    "        cleancmd = ('''\\nif [ -f whatshap/phased/{REGION}.whatshap.vcf.gz.tbi ];then '''\n",
    "                    '''rm -f bams/subset/*{REGION}*bam*;'''\n",
    "                    '''fi\\n''').format(REGION = REGION)\n",
    "        subfile.write(cleancmd)\n",
    "        subfile.close()\n",
    "        outfile.write(f\"bash cmds/{REGION}.whatshap.sh\\n\")\n",
    "        \n",
    "    ### SLURM ###\n",
    "    slurm_cmd = f\"ls cmds/{CHROM}:*whatshap.sh | sed 's/^/bash /g' | parallel --progress --jobs 10\"\n",
    "    slurm = make_slurm(echo = True, id = f\"{CHROM}.whatshap\",cmd_string = slurm_cmd, mem = '200000',time = '120:00:00', c = '10', p = 'long')\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42270758",
   "metadata": {},
   "source": [
    "Run for small interstitial regions to facilitate ligation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9fc7522a",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p whatshap whatshap/bams/ whatshap/phased whatshap/vcfs\n",
    " \n",
    "    \n",
    "AB_sizes = {}\n",
    "with open('white_abalone.fasta.fai','r') as infile:\n",
    "    for line in infile:\n",
    "        chrom, size, total, b, c = line.split('\\t')\n",
    "        if int(size) > 1e6:\n",
    "            AB_sizes[chrom] = int(size)\n",
    "infile.close()\n",
    "AB_sizes\n",
    "\n",
    "\n",
    "#### This code creates a single slurm script PER CHROMOSOME ####\n",
    "#### but parallelizes within the slurm script using subregions ####\n",
    "#### and multithreading ####\n",
    "for CHROM, SIZE in AB_sizes.items():\n",
    "    ##\n",
    "    SIZE = int(SIZE)\n",
    "    for I in range(1,SIZE,5000000):\n",
    "        \n",
    "        ### REGION SETUP AND CHECKS ###\n",
    "        START = I\n",
    "        END = I + 5000000 - 1\n",
    "        if END > SIZE:\n",
    "            END = SIZE\n",
    "        seg_start = END - 100000\n",
    "        seg_end = END + 100000\n",
    "        REGION = '%s:%s-%s' % (CHROM, seg_start, seg_end)\n",
    "        if os.path.exists(f\"whatshap/phased/{REGION}.vcf.gz.tbi\"): continue\n",
    "        baminput = ''\n",
    "        subfile = open(f\"cmds/{REGION}.inter.sh\", 'w')\n",
    "        \n",
    "        ### VCF CMD ###\n",
    "        cmd = f\"bcftools view -O z -r {REGION} vcfs/biallelic.vcf.gz > whatshap/vcfs/{REGION}.vcf.gz && tabix -f whatshap/vcfs/{REGION}.vcf.gz\\n\"\n",
    "        subfile.write(cmd)\n",
    "        \n",
    "        ### SUBBAMS CMDS ###\n",
    "        for SAMPLE in samples:\n",
    "            BAM = f\"aln/{SAMPLE}.mdup.bam\"\n",
    "            SUBBAM = f\"whatshap/bams/{SAMPLE}.{REGION}.bam\"\n",
    "            if not os.path.exists(SUBBAM + '.bai'):\n",
    "                samcmd = f\"samtools view -b -1 -M {BAM} {REGION} > {SUBBAM} && samtools index -b {SUBBAM}\\n\"\n",
    "                subfile.write(samcmd)\n",
    "            baminput += '%s ' % SUBBAM\n",
    "        \n",
    "        ### PHASE CMD ###\n",
    "        phasecmd = ('''\\nwhatshap phase '''\n",
    "                    '''--ped poplists/plink.ped '''\n",
    "                    '''whatshap/vcfs/{REGION}.vcf.gz '''\n",
    "                    '''{baminput} '''\n",
    "                    '''-o whatshap/phased/{REGION}.whatshap.vcf '''\n",
    "                    '''--reference=white_abalone.fasta '''\n",
    "                    #'''--ignore-read-groups '''\n",
    "                    '''--chromosome {CHROM} '''\n",
    "                    '''&& bgzip -f whatshap/phased/{REGION}.whatshap.vcf && tabix -f whatshap/phased/{REGION}.whatshap.vcf.gz &> logs/{REGION}.whatshap.log\\n''').format(baminput = baminput, REGION = REGION, CHROM = CHROM)\n",
    "        subfile.write(phasecmd)\n",
    "        \n",
    "        ### CLEANUP CMD ###\n",
    "        cleancmd = ('''\\nif [ -f whatshap/phased/{REGION}.whatshap.vcf.gz.tbi ];then '''\n",
    "                    '''rm -f bams/subset/*{REGION}*bam*;'''\n",
    "                    '''fi\\n''').format(REGION = REGION)\n",
    "        subfile.write(cleancmd)\n",
    "        subfile.close()\n",
    "        \n",
    "    ### SLURM ###\n",
    "    slurm_cmd = f\"ls cmds/{CHROM}:*inter.sh | sed 's/^/bash /g' | parallel --progress --jobs 10\"\n",
    "    slurm = make_slurm(echo = False, id = f\"{CHROM}.inter\",cmd_string = slurm_cmd, mem = '60000',time = '08:00:00', c = '10', p = 'long')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be1c3a7d",
   "metadata": {},
   "source": [
    "Ligate "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d436a453",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HiC_scaffold_1 done, ready to ligate\n",
      "Submitted batch job 3336728\n",
      "HiC_scaffold_2 done, ready to ligate\n",
      "Submitted batch job 3336729\n",
      "HiC_scaffold_3 done, ready to ligate\n",
      "Submitted batch job 3336730\n",
      "HiC_scaffold_4 done, ready to ligate\n",
      "Submitted batch job 3336731\n",
      "HiC_scaffold_5 done, ready to ligate\n",
      "Submitted batch job 3336732\n",
      "HiC_scaffold_6 done, ready to ligate\n",
      "Submitted batch job 3336733\n",
      "HiC_scaffold_7 done, ready to ligate\n",
      "Submitted batch job 3336734\n",
      "HiC_scaffold_8 done, ready to ligate\n",
      "Submitted batch job 3336735\n",
      "HiC_scaffold_9 done, ready to ligate\n",
      "Submitted batch job 3336736\n",
      "HiC_scaffold_10 done, ready to ligate\n",
      "Submitted batch job 3336737\n",
      "HiC_scaffold_11 done, ready to ligate\n",
      "Submitted batch job 3336738\n",
      "HiC_scaffold_12 done, ready to ligate\n",
      "Submitted batch job 3336739\n",
      "HiC_scaffold_13 done, ready to ligate\n",
      "Submitted batch job 3336740\n",
      "HiC_scaffold_14 done, ready to ligate\n",
      "Submitted batch job 3336741\n",
      "HiC_scaffold_15 done, ready to ligate\n",
      "Submitted batch job 3336742\n",
      "HiC_scaffold_16 done, ready to ligate\n",
      "Submitted batch job 3336743\n",
      "HiC_scaffold_17 done, ready to ligate\n",
      "Submitted batch job 3336744\n",
      "HiC_scaffold_18 done, ready to ligate\n",
      "Submitted batch job 3336745\n"
     ]
    }
   ],
   "source": [
    "!mkdir -p whatshap/ligated\n",
    "\n",
    "def checkinput(CHROM):\n",
    "    chunksfile = open('whatshap/%s_chunks.txt' % CHROM, 'w')\n",
    "    SIZE = int(AB_sizes[CHROM])\n",
    "    for I in range(1,SIZE,5000000):\n",
    "        START = I\n",
    "        END = I + 5000000 - 1\n",
    "        if END > SIZE:\n",
    "            END = SIZE\n",
    "            INTER_REGION = None\n",
    "        else:\n",
    "            seg_start = END - 100000\n",
    "            seg_end = END + 100000\n",
    "            INTER_REGION = '%s:%s-%s' % (CHROM, seg_start, seg_end)\n",
    "\n",
    "        REGION = '%s:%s-%s' % (CHROM, START, END)\n",
    "        \n",
    "        if not os.path.exists('whatshap/phased/%s.whatshap.vcf.gz' % REGION): \n",
    "            return\n",
    "        else:\n",
    "            chunksfile.write('whatshap/phased/%s.whatshap.vcf.gz\\n' % REGION)\n",
    "        if INTER_REGION:\n",
    "            chunksfile.write('whatshap/phased/%s.whatshap.vcf.gz\\n' % INTER_REGION)\n",
    "    \n",
    "    chunksfile.close()\n",
    "    return('done')\n",
    "\n",
    "for CHROM, SIZE in AB_sizes.items():\n",
    "    if checkinput(CHROM):\n",
    "        if os.path.exists('whatshap/ligated/%s.ligated.vcf.gz.tbi' % CHROM):\n",
    "            continue\n",
    "        print('%s done, ready to ligate' % CHROM)\n",
    "        cmd = ('''bcftools concat --ligate '''\n",
    "               '''-f whatshap/{CHROM}_chunks.txt '''\n",
    "               '''-O z > whatshap/ligated/{CHROM}.ligated.vcf.gz && tabix -f whatshap/ligated/{CHROM}.ligated.vcf.gz\\n''').format(CHROM = CHROM)\n",
    "        slurm = make_slurm(run = True, id = f\"{CHROM}.ligate\", cmd_string = cmd, mem = '5000', time = '24:00:00', p = 'long')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24522e8e",
   "metadata": {},
   "source": [
    "Concatenate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "4533958b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking the headers and starting positions of 18 files\n",
      "Concatenating whatshap/ligated/HiC_scaffold_1.ligated.vcf.gz\t11.427196 seconds\n",
      "Concatenating whatshap/ligated/HiC_scaffold_2.ligated.vcf.gz\t11.308870 seconds\n",
      "Concatenating whatshap/ligated/HiC_scaffold_3.ligated.vcf.gz\t12.645885 seconds\n",
      "Concatenating whatshap/ligated/HiC_scaffold_4.ligated.vcf.gz\t8.985294 seconds\n",
      "Concatenating whatshap/ligated/HiC_scaffold_5.ligated.vcf.gz\t12.284028 seconds\n",
      "Concatenating whatshap/ligated/HiC_scaffold_6.ligated.vcf.gz\t13.187475 seconds\n",
      "Concatenating whatshap/ligated/HiC_scaffold_7.ligated.vcf.gz\t19.020247 seconds\n",
      "Concatenating whatshap/ligated/HiC_scaffold_8.ligated.vcf.gz\t8.479397 seconds\n",
      "Concatenating whatshap/ligated/HiC_scaffold_9.ligated.vcf.gz\t10.283542 seconds\n",
      "Concatenating whatshap/ligated/HiC_scaffold_10.ligated.vcf.gz\t10.326275 seconds\n",
      "Concatenating whatshap/ligated/HiC_scaffold_11.ligated.vcf.gz\t10.409046 seconds\n",
      "Concatenating whatshap/ligated/HiC_scaffold_12.ligated.vcf.gz\t20.642290 seconds\n",
      "Concatenating whatshap/ligated/HiC_scaffold_13.ligated.vcf.gz\t10.343869 seconds\n",
      "Concatenating whatshap/ligated/HiC_scaffold_14.ligated.vcf.gz\t9.059974 seconds\n",
      "Concatenating whatshap/ligated/HiC_scaffold_15.ligated.vcf.gz\t8.418869 seconds\n",
      "Concatenating whatshap/ligated/HiC_scaffold_16.ligated.vcf.gz\t10.304036 seconds\n",
      "Concatenating whatshap/ligated/HiC_scaffold_17.ligated.vcf.gz\t10.613987 seconds\n",
      "Concatenating whatshap/ligated/HiC_scaffold_18.ligated.vcf.gz\t15.933866 seconds\n"
     ]
    }
   ],
   "source": [
    "catcmd = 'bcftools concat -O z -o whatshap/biallelic.phased.vcf.gz '\n",
    "for CHROM in AB_sizes.keys():\n",
    "    catcmd += f\"whatshap/ligated/{CHROM}.ligated.vcf.gz \"\n",
    "!$catcmd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "697dc13f",
   "metadata": {},
   "source": [
    "Cleanup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "f4979fd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "rm whatshap/phased/* whatshap/ligated/*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "076363a3-bb00-4b04-af53-50009ffbc922",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "34ba1122-4173-46b0-bfe0-bdd8379f9ab8",
   "metadata": {},
   "source": [
    "# POOHA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "484c1631-854f-4405-b969-e332ef4c7bcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_by_candidates = {}\n",
    "with open('final/all_sample_by_candidates.txt','r') as infile:\n",
    "    for line in infile:\n",
    "        parts = line.split('\\t')\n",
    "        sample,chrom,start,end = parts[0:4]\n",
    "        if sample not in sample_by_candidates.keys():\n",
    "            sample_by_candidates[sample] = [[chrom,start,end]]\n",
    "        else:\n",
    "            sample_by_candidates[sample].append([chrom,start,end])\n",
    "infile.close()\n",
    "\n",
    "sample_2_parents = {}\n",
    "with open('poplists/child_2_parent.txt','r') as infile:\n",
    "    for line in infile:\n",
    "        line = line.strip()\n",
    "        child, parentstring = line.split('\\t')\n",
    "        sample_2_parents[child] = parentstring.split(',')\n",
    "infile.close()\n",
    "\n",
    "\n",
    "!mkdir -p haplotypes\n",
    "\n",
    "allcmds = 'rm -f haplotypes/pooha_output.txt\\n'\n",
    "for SAMPLE, CANDIDATES in sample_by_candidates.items():\n",
    "    CHILD = SAMPLE\n",
    "    P0, P1 = sample_2_parents[CHILD]\n",
    "    for CAN in CANDIDATES:\n",
    "        chrom, start, end = CAN\n",
    "        left = int(start) - 50000\n",
    "        right = int(end) + 50000\n",
    "        cmd = ('''/private/groups/shapirolab/brock/Software/POOHA/POOHA '''\n",
    "               '''--verbose --min-parents-GQ 60 --min-child-GQ 60 '''\n",
    "               '''--max-marker-distance 10000 '''\n",
    "               '''--region {chrom}:{left}-{right} '''\n",
    "               '''--conflicting_pairs haplotypes/{CHILD}_{chrom}_{end}_conflicting.txt '''\n",
    "               '''--output_variants germline '''\n",
    "               '''vcfs/biallelic.vcf.gz '''\n",
    "               '''{P0} {P1} {CHILD} '''\n",
    "               '''aln/{CHILD}.mdup.bam | sed \"s/ P / {P0} /;s/ M / {P1} /\" '''\n",
    "               '''>> haplotypes/pooha_output.txt'''\n",
    "               '''\\n''').format(CHILD = CHILD, P0 = P0, P1 = P1, chrom = chrom, end = end, left = left, right = right)\n",
    "        allcmds += cmd\n",
    "slurm = make_slurm(id = f\"pooha\", cmd_string = allcmds, mem = '10000', time = '04:00:00', p = 'long')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02d96cb8-e724-43bf-9250-fff933bbe834",
   "metadata": {},
   "source": [
    "Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3585b5d4-2afe-417f-80bf-c7d59e126044",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "library(data.table)\n",
    "library(tidyverse)\n",
    "library(magrittr)\n",
    "library(ggplot2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "945e9b99-7603-4b77-8185-5a43fb77273b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Joining with `by = join_by(child, chrom, pos, ref, alt)`\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "ph = fread(cmd = 'grep -v \"chrom\" haplotypes/pooha_output.txt')\n",
    "ph %<>% set_colnames(c('chrom', 'pos', 'ref', 'alt', 'child', 'parent', 'method', 'parent_qual', 'variant_type', 'n_informative_reads', 'n_informative_vars', 'n_likely_somatic_reads', 'n_conflicting_reads' ))\n",
    "\n",
    "candidates = fread('final/all_sample_by_candidates.txt') \n",
    "candidates %<>% set_colnames(c('child','chrom','bedstart','pos','ref','alt'))\n",
    "\n",
    "cd_parents = \n",
    "    left_join(candidates, ph) %>%\n",
    "    filter(parent != 'U')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b0d9c609-1377-4e9a-88a6-5c7625ae0c2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Joining with `by = join_by(child, chrom, pos, ref, alt)`\n",
      "            child           chrom bedstart      pos    ref    alt   parent\n",
      "           <char>          <char>    <int>    <int> <char> <char>   <char>\n",
      " 1:          FG29  HiC_scaffold_2 67338091 67338092      T      A        U\n",
      " 2:          FG29  HiC_scaffold_2 71549310 71549311      G      A        U\n",
      " 3:          FG29  HiC_scaffold_3 31079938 31079939      G      C     Y121\n",
      " 4:          FG29  HiC_scaffold_6 20837340 20837341      C      G        U\n",
      " 5:          FG29  HiC_scaffold_7 33841068 33841069      C      T Green312\n",
      " 6:          FG29  HiC_scaffold_9 55177073 55177074      T      G        U\n",
      " 7:          FG29 HiC_scaffold_10  6202422  6202423      G      A        U\n",
      " 8:          FG29 HiC_scaffold_10 54470578 54470579      G      A        U\n",
      " 9:          FG29 HiC_scaffold_11 47449978 47449979      G      T     Y121\n",
      "10:          FG29 HiC_scaffold_14 31764071 31764072      G      A        U\n",
      "11:          FG29 HiC_scaffold_15 10189581 10189582      G      A        U\n",
      "12:          FG29 HiC_scaffold_15 27923854 27923855      A      T        U\n",
      "13:          FG29 HiC_scaffold_15 31631911 31631912      C      A     Y121\n",
      "14:          FG29 HiC_scaffold_17 35289449 35289450      G      T     Y121\n",
      "15:          FG30  HiC_scaffold_2 64182269 64182270      C      A        U\n",
      "16:          FG30 HiC_scaffold_13  2046468  2046469      T      A     Y121\n",
      "17:          FG30 HiC_scaffold_13 37482984 37482985      T      G        U\n",
      "18:          FG30 HiC_scaffold_16 43758122 43758123      C      A     Y121\n",
      "19:          FG30 HiC_scaffold_16 45181553 45181554      A      C        U\n",
      "20:          FG33  HiC_scaffold_1 51016392 51016393      G      T        U\n",
      "21:          FG33  HiC_scaffold_4  6842929  6842930      T      C     Y121\n",
      "22:          FG33  HiC_scaffold_4  8785696  8785697      C      T     Y121\n",
      "23:          FG33  HiC_scaffold_4  9688221  9688222      A      T        U\n",
      "24:          FG33  HiC_scaffold_5 21326718 21326719      G      C        U\n",
      "25:          FG33  HiC_scaffold_5 45739036 45739037      T      A        U\n",
      "26:          FG33  HiC_scaffold_6 41743405 41743406      G      C        U\n",
      "27:          FG33  HiC_scaffold_6 49235550 49235551      C      G     Y121\n",
      "28:          FG33  HiC_scaffold_7 31541473 31541474      G      C     Y121\n",
      "29:          FG33 HiC_scaffold_13 57947702 57947703      C      A     Y121\n",
      "30:          FG33 HiC_scaffold_16 43182378 43182379      C      T        U\n",
      "31:          FG33 HiC_scaffold_17 22164370 22164371      G      C     Y121\n",
      "32:   NorbertW740  HiC_scaffold_4 54454190 54454191      T      C        U\n",
      "33:   NorbertW740  HiC_scaffold_9 25871202 25871203      C      T   BML397\n",
      "34:   NorbertW740 HiC_scaffold_10 16544715 16544716      G      A        U\n",
      "35:   NorbertW740 HiC_scaffold_13 16568119 16568120      G      A   BML309\n",
      "36:   NorbertW740 HiC_scaffold_14 50782953 50782954      T      C        U\n",
      "37:   NorbertW740 HiC_scaffold_16 24471654 24471655      T      C        U\n",
      "38:   NorbertW756  HiC_scaffold_1 39792912 39792913      G      A        U\n",
      "39:   NorbertW756  HiC_scaffold_4  1618125  1618126      T      G        U\n",
      "40:   NorbertW756  HiC_scaffold_4  7399245  7399246      C      T        U\n",
      "41:   NorbertW756  HiC_scaffold_6 56619377 56619378      T      C        U\n",
      "42:   NorbertW756 HiC_scaffold_10 50808776 50808777      G      A        U\n",
      "43:   NorbertW756 HiC_scaffold_12 28098443 28098444      G      A        U\n",
      "44:   NorbertW756 HiC_scaffold_18  5576792  5576793      T      A        U\n",
      "45:   NorbertW802  HiC_scaffold_1 48449274 48449275      T      A        U\n",
      "46:   NorbertW802  HiC_scaffold_1 87217394 87217395      G      T   BML309\n",
      "47:   NorbertW802  HiC_scaffold_4  3018727  3018728      C      T        U\n",
      "48:   NorbertW802  HiC_scaffold_6 16441447 16441448      G      A        U\n",
      "49:   NorbertW802 HiC_scaffold_12 10406216 10406217      G      A        U\n",
      "50:   NorbertW802 HiC_scaffold_13 52062399 52062400      T      C   BML397\n",
      "51:   NorbertW802 HiC_scaffold_16 51322843 51322844      C      A        U\n",
      "52:   NorbertW853  HiC_scaffold_2 31835901 31835902      T      C        U\n",
      "53:   NorbertW853  HiC_scaffold_2 66199100 66199101      C      T        U\n",
      "54:   NorbertW853  HiC_scaffold_3  7022660  7022661      A      T        U\n",
      "55:   NorbertW853  HiC_scaffold_3 39551501 39551502      C      G        U\n",
      "56:   NorbertW853  HiC_scaffold_4 37670769 37670770      G      A        U\n",
      "57:   NorbertW853  HiC_scaffold_7 49030321 49030322      G      A   BML397\n",
      "58:   NorbertW853  HiC_scaffold_8 48129883 48129884      C      T        U\n",
      "59:   NorbertW853  HiC_scaffold_8 61265371 61265372      A      G        U\n",
      "60:   NorbertW853 HiC_scaffold_11 10151234 10151235      A      T   BML309\n",
      "61:   NorbertW853 HiC_scaffold_16  5510720  5510721      G      A        U\n",
      "62: ToothlessL824  HiC_scaffold_4 72889403 72889404      A      C        U\n",
      "63: ToothlessL824  HiC_scaffold_6 41367342 41367343      T      C   BML397\n",
      "64: ToothlessL824  HiC_scaffold_7 24369086 24369087      C      A     Y119\n",
      "65: ToothlessL824 HiC_scaffold_11 51469515 51469516      A      G   BML397\n",
      "66: ToothlessL824 HiC_scaffold_16 16565415 16565416      A      G        U\n",
      "67: ToothlessL960  HiC_scaffold_6 22617314 22617315      C      T   BML397\n",
      "68: ToothlessL960  HiC_scaffold_8 50474106 50474107      G      A     Y119\n",
      "69: ToothlessL960  HiC_scaffold_9 42092584 42092585      C      G        U\n",
      "70: ToothlessL960 HiC_scaffold_10 44646255 44646256      A      G     Y119\n",
      "71: ToothlessL960 HiC_scaffold_11 30035456 30035457      T      C        U\n",
      "72: ToothlessL960 HiC_scaffold_11 44319500 44319501      A      G   BML397\n",
      "73: ToothlessL960 HiC_scaffold_16  7894734  7894735      T      C        U\n",
      "74: ToothlessL960 HiC_scaffold_17 22058068 22058069      A      G   BML397\n",
      "75: ToothlessL960 HiC_scaffold_18 39638851 39638852      T      C     Y119\n",
      "            child           chrom bedstart      pos    ref    alt   parent\n",
      "        method parent_qual variant_type n_informative_reads n_informative_vars\n",
      "        <char>       <int>       <char>               <int>              <int>\n",
      " 1: unassigned           0   ref_denovo                   0                  0\n",
      " 2: unassigned           0   ref_denovo                   0                  0\n",
      " 3:     phased          59   ref_denovo                   1                  1\n",
      " 4: unassigned           0   ref_denovo                   0                  0\n",
      " 5:     phased         119   ref_denovo                 139                  4\n",
      " 6: unassigned           0   ref_denovo                   0                  0\n",
      " 7: unassigned           0   ref_denovo                   0                  0\n",
      " 8: unassigned           0   ref_denovo                   0                  0\n",
      " 9:     phased         966   ref_denovo                  22                  2\n",
      "10: unassigned           0   ref_denovo                   0                  0\n",
      "11: unassigned           0   ref_denovo                   0                  0\n",
      "12: unassigned           0   ref_denovo                   0                  0\n",
      "13:     phased         230   ref_denovo                  16                  2\n",
      "14:     phased         102   ref_denovo                  16                  1\n",
      "15: unassigned           0   ref_denovo                   0                  0\n",
      "16:     phased         228   ref_denovo                   8                  2\n",
      "17: unassigned           0   ref_denovo                   0                  0\n",
      "18:     phased        1000   ref_denovo                 141                  6\n",
      "19: unassigned           0   ref_denovo                   0                  0\n",
      "20: unassigned           0   ref_denovo                   0                  0\n",
      "21:     phased        1000   ref_denovo                  24                  1\n",
      "22:     phased          60   ref_denovo                   1                  1\n",
      "23: unassigned           0   ref_denovo                   0                  0\n",
      "24: unassigned           0   ref_denovo                   0                  0\n",
      "25: unassigned           0   ref_denovo                   0                  0\n",
      "26: unassigned           0   ref_denovo                   0                  0\n",
      "27:     phased        1000   ref_denovo                 280                 15\n",
      "28:     phased          91   ref_denovo                   2                  1\n",
      "29:     phased          59   ref_denovo                   1                  1\n",
      "30: unassigned           0   ref_denovo                   0                  0\n",
      "31:     phased         101   ref_denovo                   6                  1\n",
      "32: unassigned           0   ref_denovo                   0                  0\n",
      "33:     phased         111   ref_denovo                  13                  1\n",
      "34: unassigned           0   ref_denovo                   0                  0\n",
      "35:     phased         363   ref_denovo                   9                  2\n",
      "36: unassigned           0   ref_denovo                   0                  0\n",
      "37: unassigned           0   ref_denovo                   0                  0\n",
      "38: unassigned           0   ref_denovo                   0                  0\n",
      "39: unassigned           0   ref_denovo                   0                  0\n",
      "40: unassigned           0   ref_denovo                   0                  0\n",
      "41: unassigned           0   ref_denovo                   0                  0\n",
      "42: unassigned           0   ref_denovo                   0                  0\n",
      "43: unassigned           0   ref_denovo                   0                  0\n",
      "44: unassigned           0   ref_denovo                   0                  0\n",
      "45: unassigned           0   ref_denovo                   0                  0\n",
      "46:     phased         120   ref_denovo                  10                  1\n",
      "47: unassigned           0   ref_denovo                   0                  0\n",
      "48: unassigned           0   ref_denovo                   0                  0\n",
      "49: unassigned           0   ref_denovo                   0                  0\n",
      "50:     phased          59   ref_denovo                   1                  1\n",
      "51: unassigned           0   ref_denovo                   0                  0\n",
      "52: unassigned           0   ref_denovo                   0                  0\n",
      "53: unassigned           0   ref_denovo                   0                  0\n",
      "54: unassigned           0   ref_denovo                   0                  0\n",
      "55: unassigned           0   ref_denovo                   0                  0\n",
      "56: unassigned           0   ref_denovo                   0                  0\n",
      "57:     phased          59   ref_denovo                  21                  1\n",
      "58: unassigned           0   ref_denovo                   0                  0\n",
      "59: unassigned           0   ref_denovo                   0                  0\n",
      "60:     phased         591   ref_denovo                  98                  6\n",
      "61: unassigned           0   ref_denovo                   0                  0\n",
      "62: unassigned           0   ref_denovo                   0                  0\n",
      "63:     phased         118   ref_denovo                  20                  1\n",
      "64:     phased         101   ref_denovo                   2                  1\n",
      "65:     phased        1000   ref_denovo                   4                  2\n",
      "66: unassigned           0   ref_denovo                   0                  0\n",
      "67:     phased        1000   ref_denovo                  22                  1\n",
      "68:     phased         794   ref_denovo                 149                  6\n",
      "69: unassigned           0   ref_denovo                   0                  0\n",
      "70:     phased         123   ref_denovo                  68                  1\n",
      "71: unassigned           0   ref_denovo                   0                  0\n",
      "72:     phased        1000   ref_denovo                  36                  6\n",
      "73: unassigned           0   ref_denovo                   0                  0\n",
      "74:     phased         121   ref_denovo                   7                  1\n",
      "75:     phased         518   ref_denovo                  28                  7\n",
      "        method parent_qual variant_type n_informative_reads n_informative_vars\n",
      "    n_likely_somatic_reads n_conflicting_reads\n",
      "                     <int>               <int>\n",
      " 1:                      0                   0\n",
      " 2:                      0                   0\n",
      " 3:                      0                   0\n",
      " 4:                      0                   0\n",
      " 5:                      1                   3\n",
      " 6:                      0                   0\n",
      " 7:                      0                   0\n",
      " 8:                      0                   0\n",
      " 9:                      1                   0\n",
      "10:                      0                   0\n",
      "11:                      0                   0\n",
      "12:                      0                   0\n",
      "13:                      1                   0\n",
      "14:                      0                   0\n",
      "15:                      0                   0\n",
      "16:                      0                   0\n",
      "17:                      0                   0\n",
      "18:                      3                   2\n",
      "19:                      0                   0\n",
      "20:                      0                   0\n",
      "21:                      0                   1\n",
      "22:                      0                   0\n",
      "23:                      0                   0\n",
      "24:                      0                   0\n",
      "25:                      0                   0\n",
      "26:                      0                   0\n",
      "27:                      1                   5\n",
      "28:                      0                   0\n",
      "29:                      0                   0\n",
      "30:                      0                   0\n",
      "31:                      1                   0\n",
      "32:                      0                   0\n",
      "33:                      0                   0\n",
      "34:                      0                   0\n",
      "35:                      0                   0\n",
      "36:                      0                   0\n",
      "37:                      0                   0\n",
      "38:                      0                   0\n",
      "39:                      0                   0\n",
      "40:                      0                   0\n",
      "41:                      0                   0\n",
      "42:                      0                   0\n",
      "43:                      0                   0\n",
      "44:                      0                   0\n",
      "45:                      0                   0\n",
      "46:                      0                   0\n",
      "47:                      0                   0\n",
      "48:                      0                   0\n",
      "49:                      0                   0\n",
      "50:                      0                   0\n",
      "51:                      0                   0\n",
      "52:                      0                   0\n",
      "53:                      0                   0\n",
      "54:                      0                   0\n",
      "55:                      0                   0\n",
      "56:                      0                   0\n",
      "57:                      0                   0\n",
      "58:                      0                   0\n",
      "59:                      0                   0\n",
      "60:                      0                   3\n",
      "61:                      0                   0\n",
      "62:                      0                   0\n",
      "63:                      0                   1\n",
      "64:                      0                   0\n",
      "65:                      0                   1\n",
      "66:                      0                   0\n",
      "67:                      0                   1\n",
      "68:                      4                   0\n",
      "69:                      0                   0\n",
      "70:                      0                   2\n",
      "71:                      0                   0\n",
      "72:                      0                   3\n",
      "73:                      0                   0\n",
      "74:                      0                   1\n",
      "75:                      0                   1\n",
      "    n_likely_somatic_reads n_conflicting_reads\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "left_join(candidates, ph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a79d89fe-ef82-4d72-81b7-cbb48b8affc5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "            child           chrom bedstart      pos    ref    alt   parent\n",
      "           <char>          <char>    <int>    <int> <char> <char>   <char>\n",
      " 1:          FG29  HiC_scaffold_3 31079938 31079939      G      C     Y121\n",
      " 2:          FG29  HiC_scaffold_7 33841068 33841069      C      T Green312\n",
      " 3:          FG29 HiC_scaffold_11 47449978 47449979      G      T     Y121\n",
      " 4:          FG29 HiC_scaffold_15 31631911 31631912      C      A     Y121\n",
      " 5:          FG29 HiC_scaffold_17 35289449 35289450      G      T     Y121\n",
      " 6:          FG30 HiC_scaffold_13  2046468  2046469      T      A     Y121\n",
      " 7:          FG30 HiC_scaffold_16 43758122 43758123      C      A     Y121\n",
      " 8:          FG33  HiC_scaffold_4  6842929  6842930      T      C     Y121\n",
      " 9:          FG33  HiC_scaffold_4  8785696  8785697      C      T     Y121\n",
      "10:          FG33  HiC_scaffold_6 49235550 49235551      C      G     Y121\n",
      "11:          FG33  HiC_scaffold_7 31541473 31541474      G      C     Y121\n",
      "12:          FG33 HiC_scaffold_13 57947702 57947703      C      A     Y121\n",
      "13:          FG33 HiC_scaffold_17 22164370 22164371      G      C     Y121\n",
      "14:   NorbertW740  HiC_scaffold_9 25871202 25871203      C      T   BML397\n",
      "15:   NorbertW740 HiC_scaffold_13 16568119 16568120      G      A   BML309\n",
      "16:   NorbertW802  HiC_scaffold_1 87217394 87217395      G      T   BML309\n",
      "17:   NorbertW802 HiC_scaffold_13 52062399 52062400      T      C   BML397\n",
      "18:   NorbertW853  HiC_scaffold_7 49030321 49030322      G      A   BML397\n",
      "19:   NorbertW853 HiC_scaffold_11 10151234 10151235      A      T   BML309\n",
      "20: ToothlessL824  HiC_scaffold_6 41367342 41367343      T      C   BML397\n",
      "21: ToothlessL824  HiC_scaffold_7 24369086 24369087      C      A     Y119\n",
      "22: ToothlessL824 HiC_scaffold_11 51469515 51469516      A      G   BML397\n",
      "23: ToothlessL960  HiC_scaffold_6 22617314 22617315      C      T   BML397\n",
      "24: ToothlessL960  HiC_scaffold_8 50474106 50474107      G      A     Y119\n",
      "25: ToothlessL960 HiC_scaffold_10 44646255 44646256      A      G     Y119\n",
      "26: ToothlessL960 HiC_scaffold_11 44319500 44319501      A      G   BML397\n",
      "27: ToothlessL960 HiC_scaffold_17 22058068 22058069      A      G   BML397\n",
      "28: ToothlessL960 HiC_scaffold_18 39638851 39638852      T      C     Y119\n",
      "            child           chrom bedstart      pos    ref    alt   parent\n",
      "    method parent_qual variant_type n_informative_reads n_informative_vars\n",
      "    <char>       <int>       <char>               <int>              <int>\n",
      " 1: phased          59   ref_denovo                   1                  1\n",
      " 2: phased         119   ref_denovo                 139                  4\n",
      " 3: phased         966   ref_denovo                  22                  2\n",
      " 4: phased         230   ref_denovo                  16                  2\n",
      " 5: phased         102   ref_denovo                  16                  1\n",
      " 6: phased         228   ref_denovo                   8                  2\n",
      " 7: phased        1000   ref_denovo                 141                  6\n",
      " 8: phased        1000   ref_denovo                  24                  1\n",
      " 9: phased          60   ref_denovo                   1                  1\n",
      "10: phased        1000   ref_denovo                 280                 15\n",
      "11: phased          91   ref_denovo                   2                  1\n",
      "12: phased          59   ref_denovo                   1                  1\n",
      "13: phased         101   ref_denovo                   6                  1\n",
      "14: phased         111   ref_denovo                  13                  1\n",
      "15: phased         363   ref_denovo                   9                  2\n",
      "16: phased         120   ref_denovo                  10                  1\n",
      "17: phased          59   ref_denovo                   1                  1\n",
      "18: phased          59   ref_denovo                  21                  1\n",
      "19: phased         591   ref_denovo                  98                  6\n",
      "20: phased         118   ref_denovo                  20                  1\n",
      "21: phased         101   ref_denovo                   2                  1\n",
      "22: phased        1000   ref_denovo                   4                  2\n",
      "23: phased        1000   ref_denovo                  22                  1\n",
      "24: phased         794   ref_denovo                 149                  6\n",
      "25: phased         123   ref_denovo                  68                  1\n",
      "26: phased        1000   ref_denovo                  36                  6\n",
      "27: phased         121   ref_denovo                   7                  1\n",
      "28: phased         518   ref_denovo                  28                  7\n",
      "    method parent_qual variant_type n_informative_reads n_informative_vars\n",
      "    n_likely_somatic_reads n_conflicting_reads\n",
      "                     <int>               <int>\n",
      " 1:                      0                   0\n",
      " 2:                      1                   3\n",
      " 3:                      1                   0\n",
      " 4:                      1                   0\n",
      " 5:                      0                   0\n",
      " 6:                      0                   0\n",
      " 7:                      3                   2\n",
      " 8:                      0                   1\n",
      " 9:                      0                   0\n",
      "10:                      1                   5\n",
      "11:                      0                   0\n",
      "12:                      0                   0\n",
      "13:                      1                   0\n",
      "14:                      0                   0\n",
      "15:                      0                   0\n",
      "16:                      0                   0\n",
      "17:                      0                   0\n",
      "18:                      0                   0\n",
      "19:                      0                   3\n",
      "20:                      0                   1\n",
      "21:                      0                   0\n",
      "22:                      0                   1\n",
      "23:                      0                   1\n",
      "24:                      4                   0\n",
      "25:                      0                   2\n",
      "26:                      0                   3\n",
      "27:                      0                   1\n",
      "28:                      0                   1\n",
      "    n_likely_somatic_reads n_conflicting_reads\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "cd_parents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "779c7153-e882-4927-bce4-53c4d4669cbb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\t2-sample test for equality of proportions with continuity correction\n",
      "\n",
      "data:  c(13, 2) out of c(15, 15)\n",
      "X-squared = 13.333, df = 1, p-value = 0.0002607\n",
      "alternative hypothesis: two.sided\n",
      "95 percent confidence interval:\n",
      " 0.423383 1.000000\n",
      "sample estimates:\n",
      "   prop 1    prop 2 \n",
      "0.8666667 0.1333333 \n",
      "\n",
      "\n",
      "\t2-sample test for equality of proportions with continuity correction\n",
      "\n",
      "data:  c(4, 2) out of c(6, 6)\n",
      "X-squared = 0.33333, df = 1, p-value = 0.5637\n",
      "alternative hypothesis: two.sided\n",
      "95 percent confidence interval:\n",
      " -0.366768  1.000000\n",
      "sample estimates:\n",
      "   prop 1    prop 2 \n",
      "0.6666667 0.3333333 \n",
      "\n",
      "\n",
      "\t2-sample test for equality of proportions with continuity correction\n",
      "\n",
      "data:  c(6, 4) out of c(10, 10)\n",
      "X-squared = 0.2, df = 1, p-value = 0.6547\n",
      "alternative hypothesis: two.sided\n",
      "95 percent confidence interval:\n",
      " -0.3294066  0.7294066\n",
      "sample estimates:\n",
      "prop 1 prop 2 \n",
      "   0.6    0.4 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "#Patrick family\n",
    "res <- prop.test(x = c(13, 2), n = c(15, 15))\n",
    "print(res)\n",
    "\n",
    "#Norbert family\n",
    "res <- prop.test(x = c(4, 2), n = c(6, 6))\n",
    "print(res)\n",
    "\n",
    "#Toothless family\n",
    "res <- prop.test(x = c(6, 4), n = c(10, 10))\n",
    "print(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "1005e4f4-b7de-448a-b2ba-f0ae9badc8c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "            child           chrom bedstart      pos    ref    alt\n",
      "           <char>          <char>    <int>    <int> <char> <char>\n",
      " 1: ToothlessL824  HiC_scaffold_4 72889403 72889404      A      C\n",
      " 2: ToothlessL824  HiC_scaffold_6 41367342 41367343      T      C\n",
      " 3: ToothlessL824  HiC_scaffold_7 24369086 24369087      C      A\n",
      " 4: ToothlessL824 HiC_scaffold_11 51469515 51469516      A      G\n",
      " 5: ToothlessL824 HiC_scaffold_16 16565415 16565416      A      G\n",
      " 6: ToothlessL960  HiC_scaffold_6 22617314 22617315      C      T\n",
      " 7: ToothlessL960  HiC_scaffold_8 50474106 50474107      G      A\n",
      " 8: ToothlessL960  HiC_scaffold_9 42092584 42092585      C      G\n",
      " 9: ToothlessL960 HiC_scaffold_10 44646255 44646256      A      G\n",
      "10: ToothlessL960 HiC_scaffold_11 30035456 30035457      T      C\n",
      "11: ToothlessL960 HiC_scaffold_11 44319500 44319501      A      G\n",
      "12: ToothlessL960 HiC_scaffold_16  7894734  7894735      T      C\n",
      "13: ToothlessL960 HiC_scaffold_17 22058068 22058069      A      G\n",
      "14: ToothlessL960 HiC_scaffold_18 39638851 39638852      T      C\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "candidates %>% filter(grepl('Toothless',child))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "doom",
   "language": "python",
   "name": "doom"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
